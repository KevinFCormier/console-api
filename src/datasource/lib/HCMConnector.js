/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */

import lru from 'lru-cache';
import _ from 'lodash';
import request from './request';
import config from '../../../config';
import { GenericError, AuthenticationError } from './errors';

const transformResource = (clusterName, resource, resourceName) => ({
  ...resource,
  name: resourceName,
  cluster: clusterName,
});

const transform = (clusterName, resources) =>
  _.reduce(
    resources,
    (transformed, resource, resourceName) => {
      transformed.push(transformResource(clusterName, resource, resourceName));
      return transformed;
    },
    [],
  );

const clustersToItems = clusterData =>
  _.reduce(
    clusterData,
    (accum, { Results: resources }, clusterName) => {
      // Transform all resources for the cluster
      if (resources.code) {
        accum.push(resources);
      } else {
        transform(clusterName, resources).forEach(resource => accum.push(resource));
      }

      return accum;
    },
    [],
  );

export default class HCMConnector {
  constructor({
    hcmUrl = config.get('hcmUrl'),
    pollInterval = config.get('hcmPollInterval'),
    pollTimeout = config.get('hcmPollTimeout'),
    requestLib = request,
  } = {}) {
    this.hcmUrl = hcmUrl;
    this.pollInterval = pollInterval;
    this.pollTimeout = pollTimeout;
    this.request = requestLib;

    this.tokenCache = lru({
      max: 1000,
      maxAge: 60 * 60 * 1000, // 1hr
    });

    this.workDefaults = {
      method: 'POST',
      json: {
        SrcClusters: { Names: null, Labels: null, Status: null },
        DstClusters: { Names: ['*'], Labels: null, Status: ['healthy'] },
        ClientID: '',
        Dryrun: false,
        Completed: false,
        UUID: '',
        Operation: 'get',
        Work: { Namespaces: '', Status: '', Labels: '' },
        Timestamp: new Date(),
        NextRequest: null,
        FinishedRequest: null,
        Description: '',
      },
    };
  }

  async getToken(req) {
    let idToken;
    const authorization = req.headers.authorization ?
      req.headers.authorization : req.headers.Authorization;
    if (_.isEmpty(authorization) && process.env.NODE_ENV === 'development') {
      // special case for graphiql to work locally
      // do not exchange for idtoken since authorization header is empty
      idToken = 'localdev';
    } else {
      const accessToken = authorization.substring(7);
      idToken = this.tokenCache.get(accessToken);
      if (!idToken) {
        const options = {
          url: `${config.get('PLATFORM_IDENTITY_PROVIDER_URL')}/v1/auth/exchangetoken`,
          headers: {
            'Content-Type': 'application/x-www-form-urlencoded',
          },
          method: 'POST',
          json: true,
          form: {
            access_token: accessToken,
          },
        };
        const response = await this.request(options);
        idToken = _.get(response, 'body.id_token');
        if (idToken) {
          this.tokenCache.set(accessToken, idToken);
        } else {
          throw new AuthenticationError({ data: response });
        }
      }
    }
    return `Bearer ${idToken}`;
  }

  async processRequest(req, path, ...httpOverrides) {
    const opts = _.merge({
      url: `${this.hcmUrl}${path}`,
      headers: {
        Authorization: await this.getToken(req),
      },
      json: {},
      method: 'GET',
    }, ...httpOverrides);
    const { body } = await this.request(opts);
    if (body.Error) {
      throw new GenericError({ data: body.Error });
    }

    let returnValue;
    try {
      returnValue = JSON.parse(body.RetString).Result;
    } catch (err) {
      returnValue = body.RetString;
    }

    return returnValue;
  }

  timeout() {
    let timeoutID;
    const timeoutPromise = new Promise((resolve, reject) => {
      timeoutID = setTimeout(reject, this.pollTimeout, new GenericError({ data: { error: 'Manager request timed out' } }));
    });

    return { timeoutID, timeoutPromise };
  }

  poll(req, workID) {
    let intervalID;

    const pollPromise = new Promise(async (resolve, reject) => {
      // eslint-disable-next-line consistent-return
      intervalID = setInterval(async () => {
        const { Completed, Results } = await this.processRequest(req, `/api/v1alpha1/work/${workID}`, { method: 'GET' });
        if (Completed) {
          if (Results.code || Results.message) {
            return reject(Results);
          }

          resolve(Results);
        }
      }, this.pollInterval);
    });

    return { intervalID, pollPromise };
  }

  async pollWork(req, ...httpOverrides) {
    let err;
    let result;

    const workID = await this.processRequest(req, '/api/v1alpha1/work', { method: 'POST' }, ...httpOverrides);

    const { timeoutID, timeoutPromise } = this.timeout();
    const { intervalID, pollPromise } = this.poll(req, workID);

    try {
      result = await Promise.race([timeoutPromise, pollPromise]);
    } catch (error) {
      err = error;
    }

    // Clear interval/timeout in all cases - node doesn't have finally block
    clearInterval(intervalID);
    clearTimeout(timeoutID);

    if (err) {
      throw err;
    }

    return result;
  }

  async getWork(req, type, opts) {
    const result = await this.pollWork(req, { json: { Resource: type } }, this.workDefaults, opts);
    return clustersToItems(result);
  }
}
