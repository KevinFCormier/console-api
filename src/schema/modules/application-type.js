/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */

import { applications } from '../../datasource/hcm';

export const typeDef = `
# HCM Application
type Application {
  Name: String
  Annotations: JSON
  Labels: JSON
  Components: [AppNode]
  Dependencies: [AppNode]
}

# HCM Application Node (AppNode)
type AppNode {
  Name: String
  Cluster: String
  Kind: String
  Status: String
}
`;

export const applicationResolver = {
  Query: {
    applications,
  },
};
