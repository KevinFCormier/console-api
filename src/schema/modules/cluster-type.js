/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */

import { clusters } from '../../datasource/hcm';

export const typeDef = `
type Cluster {
  AgentEndpoint: String
  ClusterEndpoint: String
  ClusterName: String
  Conflict: Boolean
  Labels: JSON
  ProxyEndpoint: String
  Status: String
  TotalCpus: Int
  TotalDeployments: Int
  TotalMemory: String
  TotalNodes: Int
  TotalPods: Int
  TotalServices: Int
  TotalStorage: String
}
`;

export const clusterResolver = {
  Query: {
    clusters,
  },
};
