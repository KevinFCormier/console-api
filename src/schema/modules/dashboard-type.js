/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */

import _ from 'lodash';
import { clusters, releases, pods } from '../../datasource/hcm';

export const typeDef = `
  type TableRow {
    status: String
    link: String
    resourceName: String
    percentage: Int
  }

  type DashboardCardItem {
    name: String
    healthy: Int
    critical: Int
    warning: Int
    table: [TableRow]
    error: String
  } 
  
  type DashboardChartItem {
    name: String
    # return something looks like [ [[value1, value2, value3], title1] , [[valueA, valueB, valueC], title2] ]
    # issue created against Carbon: https://github.com/carbon-design-system/carbon-addons-data-viz-react/issues/112
    data: [[String]]
    error: String
  }
  
  type DashboardData {
    cardItems: [DashboardCardItem]
    pieChartItems: [DashboardChartItem]
  }
`;

const sortTable = table => _.sortBy(table, [(row) => {
  switch (row.status) {
    case 'critical':
      return 1;

    case 'warning':
      return 2;

    default:
      return 3;
  }
}]);

const sortCards = dashboardCards => _.sortBy(dashboardCards, [(card) => {
  switch (true) {
    case card.critical > 0:
      return -card.critical * 100;

    case card.warning > 0:
      return -card.warning * 1;

    default:
      return 0;
  }
}]);

const genericStatus = (resource) => {
  if (resource.Status) {
    switch (resource.Status.toLowerCase()) {
      case 'failed':
        return 'critical';

      // TODO: Return warning status - 06/11/18 09:35:53 sidney.wijngaarde1@ibm.com

      case 'healthy':
      case 'deployed':
        return 'healthy';

      default:
        return 'critical';
    }
  } else if (resource.State != null) {
    return resource.State ? 'healthy' : 'critical';
  } else {
    return 'critical';
  }
};

const percentageStatus = field => (resource) => {
  const percent = resource[field];
  if (!percent) {
    return 'critical';
  }

  switch (true) {
    case percent > 90:
      return 'critical';

    case percent > 75:
      return 'warning';

    default:
      return 'healthy';
  }
};

function getDashboardCard({
  name, rawData, transform, status = genericStatus,
}) {
  const cardData = rawData.reduce((accum, curr) => {
    const stat = status(curr);

    accum[stat] += 1;

    accum.table.push(transform(curr, stat));
    return accum;
  }, {
    name, healthy: 0, critical: 0, warning: 0, table: [], error: null,
  });

  cardData.table = sortTable(cardData.table).slice(0, 5);
  return cardData;
}

function getDashboardPieChart({
  name, rawData, transform, status = genericStatus,
}) {
  const chartData = rawData.reduce((accum, curr) => {
    const stat = status(curr);
    accum.data = (transform(curr, stat, accum.data));
    return accum;
  }, {
    data: {},
  });
  const result = [];
  // dirty code for supporting Carbon data schema
  // issue created against Carbon: https://github.com/carbon-design-system/carbon-addons-data-viz-react/issues/112
  // update this part once Carbon fixed the issue
  Object.keys(chartData.data).forEach((key) => {
    const tempResult = [];
    tempResult.push(key);
    tempResult.push(chartData.data[key].toString());
    result.push(tempResult);
  });
  return { name, data: result };
}

async function getDashboardItems({
  query, req, cards = [], pieCharts = [],
}) {
  try {
    const rawData = await query(null, null, req);
    const cardsMap = cards.map(card => getDashboardCard({ rawData, ...card }));
    const pieChartsItems = pieCharts.map(chart => getDashboardPieChart({ rawData, ...chart }));
    return {
      cardsMap,
      pieChartsItems,
    };
  } catch (error) {
    return cards.map(() => ({ error }));
  }
}

const transformTotalCluster = (curr, status, currentData) => {
  const currentMap = currentData;
  if (Object.prototype.hasOwnProperty.call(currentMap, status)) {
    currentMap[status] += 1;
  } else {
    currentMap[status] = 1;
  }
  return currentMap;
};

const transformCluster = (cluster, status) => ({
  link: cluster.Labels.clusterip,
  percentage: cluster.Percentage || 0,
  resourceName: cluster.ClusterName,
  status,
});

const transformRelease = (release, status) => ({
  link: '', // TODO: save cluster ip's and generate links - 06/06/18 13:12:48 sidney.wijngaarde1@ibm.com
  percentage: 0,
  resourceName: release.name,
  status,
});

const transformPod = (pod, status) => ({
  link: '',
  percentage: 0,
  resourceName: pod.name,
  status,
});

const transformPercentage = field => (cluster, status) => ({
  link: cluster.Labels.clusterip,
  percentage: Math.round(cluster[field]),
  resourceName: cluster.ClusterName,
  status,
});

export const dashboardResolver = {
  Query: {
    dashboard: async (root, args, req) => {
      const dashboardItems = await Promise.all([
        getDashboardItems({
          cards: [
            {
              name: 'clusters',
              transform: transformCluster,
            },
            {
              name: 'cpu',
              transform: transformPercentage('CPURequestsFraction'),
              status: percentageStatus('CPURequestsFraction'),
            },
            {
              name: 'memory',
              transform: transformPercentage('MemoryRequestsFraction'),
              status: percentageStatus('MemoryRequestsFraction'),
            },
            {
              name: 'storage',
              transform: transformPercentage('StorageUsageFraction'),
              status: percentageStatus('StorageUsageFraction'),
            },
          ],
          pieCharts: [
            {
              name: 'totalClusterHealth',
              transform: transformTotalCluster,
            },
          ],
          query: clusters,
          req,
        }),
        getDashboardItems({
          cards: [
            { name: 'helm releases', transform: transformRelease },
          ],
          query: releases,
          req,
        }),
        getDashboardItems({
          cards: [
            { name: 'pods', transform: transformPod },
          ],
          query: pods,
          req,
        }),
      ]);
      let allCards = [];
      let pieChartItems = [];
      dashboardItems.forEach((result) => {
        allCards = [...result.cardsMap, ...allCards];
        pieChartItems = [...result.pieChartsItems, ...pieChartItems];
      });
      return {
        cardItems: sortCards(_.flatten(allCards)),
        pieChartItems,
      };
    },
  },
};
