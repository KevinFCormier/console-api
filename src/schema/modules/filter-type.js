/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */

import { clusters } from '../../datasource/hcm';


export const typeDef = `
type Filters {
  clusterLabels: [FilterItem]
  clusterNames: [FilterItem]
}

type FilterItem {
  name: String
  id: String
  type: String
  key: String
  value: String
}


input FilterItemInput {
  type: String
  key: String
  value: String
}

input Filter {
  # used by topology
  cluster: [String]
  label: [LabelInput]
  namespace: [String]
  type: [String]
  
  # used by cluster/pods/healm releases
  # combined with cluster labels and cluster names
  resourceFilter: [FilterItemInput]
}
`;

export const transformFilters = (input) => {
  // may need to change for api-v2
  let clusterFilter = {};
  if (input && input.filter && input.filter) {
    const { resourceFilter = [] } = input.filter;
    const clusterName = [];
    const labelFilters = {};
    resourceFilter.forEach((label) => {
      const { key, value, type } = label;
      switch (type) {
        case 'clusterName':
          clusterName.push(value);
          break;
        case 'clusterLabel':
          labelFilters[key] = value;
          break;
        default:
          break;
      }
    });
    clusterFilter = {
      Labels: labelFilters, Names: clusterName,
    };
  }
  return clusterFilter;
};


export const filtersResolver = {
  Query: {
    filters: async (root, args, req) => {
      let clusterLabels = [];
      let clusterNames = [];
      const result = await clusters(null, null, req);
      result.forEach((element) => {
        if (element.Labels) {
          Object.entries(element.Labels).forEach(([key, value]) => {
            const name = `${key}=${value}`;
            if (!clusterLabels.find(n => n.name === name)) {
              clusterLabels = [...new Set([...clusterLabels, {
                id: name, name, type: 'clusterLabel', key, value,
              }])];
            }
          });
        }
        if (element.ClusterName) {
          clusterNames = [...clusterNames,
            {
              id: element.ClusterName,
              name: `cluster=${element.ClusterName}`,
              type: 'clusterName',
              value: element.ClusterName,
            }];
        }
      });
      return { clusterNames, clusterLabels };
    },
  },
};
