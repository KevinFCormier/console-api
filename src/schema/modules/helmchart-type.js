/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */

import { installHelmChart, charts } from '../../datasource/hcm';

export const typeDef = `
input InstallHelmChartInput {
  ChartName: String!
  Version: String
  RepoName: String!
  DstClusters: DestinationClusters
  ReleaseName: String
  Namespace: String
  URL: String
}

input DestinationClusters {
  Names: [String],
  Labels: [String],
  Status: [String],
  SortBy: String,
  TargetNum: Int
}

type HelmChart {
  RepoName: String
  Name: String
  Version: String
  URLs: [String]
}

type HelmChartResponse {
  ChartName: String
  ChartVersion: String
  HDetails: HelmChartDetails
  HelmName: String
  Namespace: String
  Status: String
  Version: String
  name: String
  cluster: String
}

type HelmChartDetails {
  Description: String
  FirstDeployed: Time
  Images: [ImageDetails]
  LastDeployed: Time
  Manifest: String
}

type Time {
  nanos: Int
  seconds: Int
}

type ImageDetails {
  ImageName: String,
  ImagePullPolicy: String
}
`;

export const helmChartResolver = {
  Query: { charts },
  Mutation: {
    installHelmChart: (root, { input }, req) => installHelmChart(req, input),
  },
};
