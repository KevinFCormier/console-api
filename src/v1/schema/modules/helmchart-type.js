/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */

import _ from 'lodash';

export const typeDef = `
input InstallHelmChartInput {
  ChartName: String!
  Version: String
  RepoName: String!
  DstClusters: DestinationClusters
  ReleaseName: String
  Namespace: String
  URL: String
}

input DestinationClusters {
  Names: [String]
  Labels: [String]
  Status: [String]
  SortBy: String
  TargetNum: Int
}

type HelmChart {
  RepoName: String
  Name: String
  Version: String
  URLs: [String]
}

type HelmChartResponse {
  ChartName: String
  ChartVersion: String
  code: Int
  message: String
  HDetails: HelmChartDetails
  HelmName: String
  Namespace: String
  Status: String
  Version: String
  name: String
  cluster: String
}

type HelmChartDetails {
  Description: String
  FirstDeployed: Time
  Images: [ImageDetails]
  LastDeployed: Time
  Manifest: String
}

type Time {
  nanos: Int
  seconds: Int
}

type ImageDetails {
  ImageName: String,
  ImagePullPolicy: String
}
`;

export const helmChartResolver = {
  Query: {
    charts: async (root, args, { req, hcmConnector }) => {
      const response = await hcmConnector.processRequest(req, '/api/v1alpha1/repo/*', { Resource: 'repo', Operation: 'get' });
      const helmRepos = response ? Object.values(response) : [];

      let catalog = [];
      await Promise.all(helmRepos.map(repo => hcmConnector.search(req, 'repo', repo.Name).catch(err => console.log(err))))
        .then((values) => {
          catalog = values.filter(item => item !== undefined);
        }).catch(err => console.log(err));

      const helmCharts = _.flatten(catalog.map(chart => Object.values(chart)));
      return _.sortBy(helmCharts, chart => `${chart.RepoName}/${chart.Name}`);
    },
  },
  Mutation: {
    installHelmChart: (root, { input }, { req, hcmConnector }) => {
      const response = hcmConnector.getWork(req, 'helmrels', {
        method: 'POST',
        json: {
          Operation: 'install',
          DstClusters: input.DstClusters,
          Work: {
            ChartName: `${input.RepoName}/${input.ChartName}`,
            Namespace: input.Namespace,
            ReleaseName: input.ReleaseName,
            URL: input.URL,
            Values: input.Values,
            Version: input.Version,
          },
        },
      }, true);
      return response;
    },
  },
};
