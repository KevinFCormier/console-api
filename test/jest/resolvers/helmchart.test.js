/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */

import { helmChartResolver } from '../../../src/v1/schema/modules/helmchart-type';
import HCMConnector from '../../../src/v1/datasource/lib/hcm-connector';

describe('Helm charts Resolver', () => {
  test('Correctly resolves helm charts query', async () => {
    const mockRequestLib = jest.fn();
    mockRequestLib.mockReturnValue({
      body: {
        RetString: '{"Result":{"acs-engine-autoscaler":{"RepoName":"test","Name":"acs-engine-autoscaler","Version":"2.2.0","URLs":["https://kubernetes-charts.storage.googleapis.com/acs-engine-autoscaler-2.2.0.tgz"]},"aerospike":{"RepoName":"test","Name":"aerospike","Version":"0.1.7","URLs":["https://kubernetes-charts.storage.googleapis.com/aerospike-0.1.7.tgz"]},"anchore-engine":{"RepoName":"test","Name":"anchore-engine","Version":"0.1.7","URLs":["https://kubernetes-charts.storage.googleapis.com/anchore-engine-0.1.7.tgz"]}},"Error":null}',
        Error: null,
      },
    });

    const context = {
      hcmConnector: new HCMConnector({ requestLib: mockRequestLib }),
      req: { headers: { authorization: '' } },
    };

    const response = await helmChartResolver.Query.charts(null, null, context);
    expect(mockRequestLib).toHaveBeenCalled();
    expect(response).toMatchSnapshot();
  });

  test('Correctly resolves install helm chart mutation', async () => {
    const mockRequestLib = jest.fn();
    mockRequestLib.mockReturnValue({
      body: {
        RetString: '{"Result":{"Request":{"SrcClusters":{"Names":null,"Labels":null,"Status":null,"SortBy":"","TargetNum":0},"DstClusters":{"Names":["crucial-owl"],"Labels":null,"Status":["healthy"],"SortBy":"","TargetNum":0},"ClientID":"","Dryrun":false,"Completed":false,"UUID":"21c83170-2b4c-48aa-8f22-aff9d858e59a","Resource":"helmrels","Operation":"install","Work":{"ChartName":"default/hcm-demoaccessories","Dryrun":false,"Namespace":"default","ReleaseName":"","URL":"http://9.42.80.212:8878/hcm-demoaccessories-0.1.0.tgz","Values":null,"ValuesFile":"","Version":""},"Timestamp":"2018-07-10T16:18:39.91Z","NextRequest":null,"RootRequest":null,"FinishedRequest":{"ClientID":"","Completed":false,"Description":"","Dryrun":false,"DstClusters":{"Labels":null,"Names":["crucial-owl"],"SortBy":"","Status":["healthy"],"TargetNum":0},"FinishedRequest":null,"NextRequest":null,"Operation":"install","Resource":"helmrels","RootRequest":null,"SrcClusters":{"Labels":null,"Names":null,"SortBy":"","Status":null,"TargetNum":0},"Timestamp":"2018-07-10T16:18:39.91Z","UUID":"21c83170-2b4c-48aa-8f22-aff9d858e59a","Work":{"ChartName":"default/hcm-demoaccessories","Dryrun":false,"Namespace":"default","ReleaseName":"","URL":"http://9.42.80.212:8878/hcm-demoaccessories-0.1.0.tgz","Values":null,"ValuesFile":"","Version":""}},"Description":""},"Completed":true,"Results":{"crucial-owl":{"Request":{"SrcClusters":{"Names":null,"Labels":null,"Status":null,"SortBy":"","TargetNum":0},"DstClusters":{"Names":["crucial-owl"],"Labels":null,"Status":["healthy"],"SortBy":"","TargetNum":0},"ClientID":"","Dryrun":false,"Completed":false,"UUID":"21c83170-2b4c-48aa-8f22-aff9d858e59a","Resource":"helmrels","Operation":"install","Work":{"ChartName":"default/hcm-demoaccessories","Dryrun":false,"Namespace":"default","ReleaseName":"","URL":"http://9.42.80.212:8878/hcm-demoaccessories-0.1.0.tgz","Values":null,"ValuesFile":"","Version":""},"Timestamp":"2018-07-10T16:18:39.91Z","NextRequest":null,"RootRequest":null,"FinishedRequest":{"ClientID":"","Completed":false,"Description":"","Dryrun":false,"DstClusters":{"Labels":null,"Names":["crucial-owl"],"SortBy":"","Status":["healthy"],"TargetNum":0},"FinishedRequest":null,"NextRequest":null,"Operation":"install","Resource":"helmrels","RootRequest":null,"SrcClusters":{"Labels":null,"Names":null,"SortBy":"","Status":null,"TargetNum":0},"Timestamp":"2018-07-10T16:18:39.91Z","UUID":"21c83170-2b4c-48aa-8f22-aff9d858e59a","Work":{"ChartName":"default/hcm-demoaccessories","Dryrun":false,"Namespace":"default","ReleaseName":"","URL":"http://9.42.80.212:8878/hcm-demoaccessories-0.1.0.tgz","Values":null,"ValuesFile":"","Version":""}},"Description":""},"Completed":true,"Results":{"binging-lobster":{"ChartName":"hcm-demoaccessories","ChartVersion":"0.1.0","HDetails":{"Description":"Install complete","FirstDeployed":{"nanos":619138559,"seconds":1531239519},"Images":[{"ImageName":"neo4j:latest","ImagePullPolicy":"IfNotPresent"},{"ImageName":"neo4j:latest","ImagePullPolicy":"IfNotPresent"}],"LastDeployed":{"nanos":619138559,"seconds":1531239519},"Values":"consul:\\n  enabled: false\\ninstancedb:\\n  enabled: true\\npromethues:\\n  enabled: false\\ntemplatedb:\\n  enabled: true\\n"},"HelmName":"binging-lobster","Namespace":"default","Status":"DEPLOYED","Version":1}},"Timestamp":"2018-07-10T16:18:39.907337343Z"},"myminikube":{"Request":{"SrcClusters":{"Names":null,"Labels":null,"Status":null,"SortBy":"","TargetNum":0},"DstClusters":{"Names":["crucial-owl"],"Labels":null,"Status":["healthy"],"SortBy":"","TargetNum":0},"ClientID":"","Dryrun":false,"Completed":false,"UUID":"21c83170-2b4c-48aa-8f22-aff9d858e59a","Resource":"helmrels","Operation":"install","Work":{"ChartName":"default/hcm-demoaccessories","Dryrun":false,"Namespace":"default","ReleaseName":"","URL":"http://9.42.80.212:8878/hcm-demoaccessories-0.1.0.tgz","Values":null,"ValuesFile":"","Version":""},"Timestamp":"2018-07-10T16:18:39.91Z","NextRequest":null,"RootRequest":null,"FinishedRequest":{"ClientID":"","Completed":false,"Description":"","Dryrun":false,"DstClusters":{"Labels":null,"Names":["crucial-owl"],"SortBy":"","Status":["healthy"],"TargetNum":0},"FinishedRequest":null,"NextRequest":null,"Operation":"install","Resource":"helmrels","RootRequest":null,"SrcClusters":{"Labels":null,"Names":null,"SortBy":"","Status":null,"TargetNum":0},"Timestamp":"2018-07-10T16:18:39.91Z","UUID":"21c83170-2b4c-48aa-8f22-aff9d858e59a","Work":{"ChartName":"default/hcm-demoaccessories","Dryrun":false,"Namespace":"default","ReleaseName":"","URL":"http://9.42.80.212:8878/hcm-demoaccessories-0.1.0.tgz","Values":null,"ValuesFile":"","Version":""}},"Description":""},"Completed":true,"Results":{"exiled-anteater":{"ChartName":"hcm-demoaccessories","ChartVersion":"0.1.0","HDetails":{"Description":"Install complete","FirstDeployed":{"nanos":128814404,"seconds":1531239520},"Images":[{"ImageName":"neo4j:latest","ImagePullPolicy":"IfNotPresent"},{"ImageName":"neo4j:latest","ImagePullPolicy":"IfNotPresent"}],"LastDeployed":{"nanos":128814404,"seconds":1531239520},"Values":"consul:\\n  enabled: false\\ninstancedb:\\n  enabled: true\\npromethues:\\n  enabled: false\\ntemplatedb:\\n  enabled: true\\n"},"HelmName":"exiled-anteater","Namespace":"default","Status":"DEPLOYED","Version":1}},"Timestamp":"2018-07-10T16:18:40.51696839Z"}},"Timestamp":"2018-07-10T16:18:40.51696839Z"},"Error":null}',
        Error: null,
      },
    });

    const context = {
      hcmConnector: new HCMConnector({ requestLib: mockRequestLib }),
      req: { headers: { authorization: '' } },
    };

    const inputs = {
      input: {
        DstClusters: {},
        Names: '',
        ChartName: 'hcm-demoaccessories',
        RepoName: 'default',
        Namespaces: '',
        URL: '',
        Values: '',
        Version: '',
      },
    };

    const response = await helmChartResolver.Mutation.installHelmChart(null, inputs, context);
    expect(mockRequestLib).toHaveBeenCalled();
    expect(response).toMatchSnapshot();
  });
});
