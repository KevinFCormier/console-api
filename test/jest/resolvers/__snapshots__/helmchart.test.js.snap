// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Helm charts Resolver Correctly resolves helm charts query 1`] = `
Array [
  Object {
    "Name": "acs-engine-autoscaler",
    "RepoName": "test",
    "URLs": Array [
      "https://kubernetes-charts.storage.googleapis.com/acs-engine-autoscaler-2.2.0.tgz",
    ],
    "Version": "2.2.0",
  },
  Object {
    "Name": "acs-engine-autoscaler",
    "RepoName": "test",
    "URLs": Array [
      "https://kubernetes-charts.storage.googleapis.com/acs-engine-autoscaler-2.2.0.tgz",
    ],
    "Version": "2.2.0",
  },
  Object {
    "Name": "acs-engine-autoscaler",
    "RepoName": "test",
    "URLs": Array [
      "https://kubernetes-charts.storage.googleapis.com/acs-engine-autoscaler-2.2.0.tgz",
    ],
    "Version": "2.2.0",
  },
  Object {
    "Name": "aerospike",
    "RepoName": "test",
    "URLs": Array [
      "https://kubernetes-charts.storage.googleapis.com/aerospike-0.1.7.tgz",
    ],
    "Version": "0.1.7",
  },
  Object {
    "Name": "aerospike",
    "RepoName": "test",
    "URLs": Array [
      "https://kubernetes-charts.storage.googleapis.com/aerospike-0.1.7.tgz",
    ],
    "Version": "0.1.7",
  },
  Object {
    "Name": "aerospike",
    "RepoName": "test",
    "URLs": Array [
      "https://kubernetes-charts.storage.googleapis.com/aerospike-0.1.7.tgz",
    ],
    "Version": "0.1.7",
  },
  Object {
    "Name": "anchore-engine",
    "RepoName": "test",
    "URLs": Array [
      "https://kubernetes-charts.storage.googleapis.com/anchore-engine-0.1.7.tgz",
    ],
    "Version": "0.1.7",
  },
  Object {
    "Name": "anchore-engine",
    "RepoName": "test",
    "URLs": Array [
      "https://kubernetes-charts.storage.googleapis.com/anchore-engine-0.1.7.tgz",
    ],
    "Version": "0.1.7",
  },
  Object {
    "Name": "anchore-engine",
    "RepoName": "test",
    "URLs": Array [
      "https://kubernetes-charts.storage.googleapis.com/anchore-engine-0.1.7.tgz",
    ],
    "Version": "0.1.7",
  },
]
`;

exports[`Helm charts Resolver Correctly resolves install helm chart mutation 1`] = `
Array [
  Object {
    "ChartName": "hcm-demoaccessories",
    "ChartVersion": "0.1.0",
    "HDetails": Object {
      "Description": "Install complete",
      "FirstDeployed": Object {
        "nanos": 619138559,
        "seconds": 1531239519,
      },
      "Images": Array [
        Object {
          "ImageName": "neo4j:latest",
          "ImagePullPolicy": "IfNotPresent",
        },
        Object {
          "ImageName": "neo4j:latest",
          "ImagePullPolicy": "IfNotPresent",
        },
      ],
      "LastDeployed": Object {
        "nanos": 619138559,
        "seconds": 1531239519,
      },
      "Values": "consul:
  enabled: false
instancedb:
  enabled: true
promethues:
  enabled: false
templatedb:
  enabled: true
",
    },
    "HelmName": "binging-lobster",
    "Namespace": "default",
    "Status": "DEPLOYED",
    "Version": 1,
    "cluster": "crucial-owl",
    "name": "binging-lobster",
  },
  Object {
    "ChartName": "hcm-demoaccessories",
    "ChartVersion": "0.1.0",
    "HDetails": Object {
      "Description": "Install complete",
      "FirstDeployed": Object {
        "nanos": 128814404,
        "seconds": 1531239520,
      },
      "Images": Array [
        Object {
          "ImageName": "neo4j:latest",
          "ImagePullPolicy": "IfNotPresent",
        },
        Object {
          "ImageName": "neo4j:latest",
          "ImagePullPolicy": "IfNotPresent",
        },
      ],
      "LastDeployed": Object {
        "nanos": 128814404,
        "seconds": 1531239520,
      },
      "Values": "consul:
  enabled: false
instancedb:
  enabled: true
promethues:
  enabled: false
templatedb:
  enabled: true
",
    },
    "HelmName": "exiled-anteater",
    "Namespace": "default",
    "Status": "DEPLOYED",
    "Version": 1,
    "cluster": "myminikube",
    "name": "exiled-anteater",
  },
]
`;
